File: c/LinkedListCycleII.c
// https://leetcode.com/problems/linked-list-cycle-ii/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *detectCycle(struct ListNode *head) {
    if (head == NULL || head->next == NULL)
        return NULL;
    
    struct ListNode *slow = head, *fast = head->next;
    struct ListNode *checker = NULL;
    
    while (fast->next != NULL && fast->next->next != NULL) {
        if (fast == slow) {
            if (checker == NULL)
                checker = head;
            else
                checker = checker->next;
        }
        if (checker == slow)
            return checker;
        slow = slow->next;
        fast = fast->next->next;
    }
    return NULL;
}

File: swift/LinkedListCycleII.swift
// https://leetcode.com/problems/linked-list-cycle-ii/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func detectCycle(_ head: ListNode?) -> ListNode? {
       var pointer: Set<UnsafeMutableRawPointer> = []
       var ptr : ListNode? = head
       while ptr != nil {
           let res = pointer.insert(Unmanaged.passUnretained(ptr!).toOpaque())
           if !res.inserted { return ptr }
           ptr = ptr?.next
       }
       return nil
    }
}